server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java

```diff
@@ -0,0 +1,121 @@
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.AdjustableSemaphore;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.common.util.concurrent.RunOnce;
+import org.elasticsearch.tasks.Task;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class guards the amount of stats requests a node can concurrently coordinate.
+ */
+public class StatsRequestLimiter {
+
+    public static final Setting<Integer> MAX_CONCURRENT_STATS_REQUESTS_PER_NODE = Setting.intSetting(
+        "node.stats.max_concurrent_requests",
+        100,
+        1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final AdjustableSemaphore maxConcurrentStatsRequestsPerNodeSemaphore;
+    private final Map<String, StatsHolder> stats = new ConcurrentHashMap<>();
+
+    public StatsRequestLimiter(Settings settings, ClusterSettings clusterSettings) {
+        this.maxConcurrentStatsRequestsPerNodeSemaphore = new AdjustableSemaphore(
+            MAX_CONCURRENT_STATS_REQUESTS_PER_NODE.get(settings),
+            false
+        );
+        clusterSettings.addSettingsUpdateConsumer(MAX_CONCURRENT_STATS_REQUESTS_PER_NODE, this::setMaxConcurrentStatsRequestsPerNode);
+    }
+
+    private void setMaxConcurrentStatsRequestsPerNode(int maxConcurrentBoundedDiagnosticRequestsPerNode) {
+        this.maxConcurrentStatsRequestsPerNodeSemaphore.setMaxPermits(maxConcurrentBoundedDiagnosticRequestsPerNode);
+    }
+
+    /**
+     * Checks if executing the action will remain within the limits of the max concurrent requests the node can handle. If the limit is
+     * respected the action will be executed otherwise it will throw an EsRejectedExecutionException. The method keeps track of current,
+     * completed and rejected requests per action type.
+     */
+    public <Request, Response> void tryToExecute(
+        Task task,
+        Request request,
+        ActionListener<Response> listener,
+        TriConsumer<Task, Request, ActionListener<Response>> executeAction
+    ) {
+        StatsHolder statsHolder = stats.computeIfAbsent(task.getAction(), ignored -> new StatsHolder(task.getAction()));
+        if (tryAcquire()) {
+            statsHolder.current.inc();
+            final Runnable release = new RunOnce(() -> {
+                release();
+                statsHolder.current.dec();
+                statsHolder.completed.inc();
+            });
+            boolean success = false;
+            try {
+                executeAction.apply(task, request, ActionListener.runBefore(listener, release::run));
+                success = true;
+            } finally {
+                if (success == false) {
+                    release.run();
+                }
```

Comment:
I *think* (but I'm not sure) this can double-invoke the `release.run()`.

Since the request wraps the `release::run` *before* the original listener, the `executeAction` can run, the first invocation of `release.run()` can happen (the one from `runBefore`), then the original listener can throw an exception, which would cause the `success` boolean not to be updated, and then the `release.run()` in the `finally` block would run a second time.

I guess it depends on whether ActionListeners are expected to ever throw exceptions (@DaveCTurner can probably clarify on this point), but if they can, then maybe the `runBefore` call could do something like `() -> { success.getAndSet(true); release.run(); }` and then it wouldn't have the potential for invoking twice?

Reply:
I am processing the possibilities but I do not have an answer yet. Coming asap

Reply:
Yes I believe it's possible that `executeAction` might throw an exception but still go on to complete its listener. I mean it probably shouldn't, but in general we're ok with completing an `ActionListener` multiple times so we don't protect against this.

However here the `release` runnable is a `RunOnce` so it doesn't matter.

Reply:
I will start with the suggestion that the `runBefore()` does something like this `() -> { success.getAndSet(true); release.run(); }` because I feel more confident about my analysis :). I believe this will cause double invocation in almost all cases, because:

1. `executeAction` introduces the `runBefore` to run before the original action listener starts whatever async needs to be run and finishes
2. Chances are by this point, the async work hasn't finished yet, which means the `success` is false and the `finally` will run the `release.run()` effectively releasing the semaphore while the node is still coordinating
3. When the async work finishes, the original listener will be notified and the `release.run()` will be called again.

Now let's go to the trickier part, it's not clear to me yet where could the original listener throw an exception. If I look at the code of the `RunBeforeActionListener`:
```
        @Override
        public void onResponse(T response) {
            try {
                runBefore.run();
            } catch (Exception ex) {
                super.onFailure(ex);
                return;
            }
            delegate.onResponse(response);
        }

```

Similar the failure is being handled,  the `release` will be called even if the original listener has an error. I would expect that the only situation where this can go wrong, is if the listener is never completed and the `executeAction` finished. But this effectively would mean that the node is still coordinating so we have a different problem then right?

Did I address your concern properly? Or am I still missing something?

Reply:
I think there's no need for any changes in this area - `RunOnce` already does what is needed.

It's not just about completing the listener once but also releasing things in the `finally` block, you also need to protect against completing the listener multiple times. But we already do the right thing here.

Reply:
Yep, totally didn't notice that the `release` was a `RunOnce`, that's fine then and no worries about this.